HARK/ConsumptionSaving/ConsIndShockModel_AgentSolve.py

```diff
@@ -0,0 +1,2377 @@
+    completed_cycles : integer
+        The number of cycles of the model solved before this call
+    solveMethod : str, optional
+        The name of the solution method to use, e.g. 'EGM'
+    solverType : str, optional
+        The name of the type of solver ('HARK', 'Dolo')
+    eventTiming : str, optional
+        Clarifies timing of events whose timing might otherwise be ambiguous
+    messaging_level : int, optional
+        Controls the amount of information returned to user. Varies by model.
+    """
+
+    def __init__(self, *args,
+                 stge_kind=None,
+                 parameters_solver=None,
+                 completed_cycles=0,
+                 **kwds):
+        if stge_kind is None:
+            stge_kind = {'iter_status': 'not initialized'}
+        self.E_Next_ = Nexspectations()  # Next given this period/stage choices
+        self.Ante_E_ = Prospectations()  # Before this stage shocks/choices
+        self.Pars = Parameters()
+        self.Bilt = Built()
+        self.Bilt.completed_cycles = completed_cycles
+        self.Bilt.stge_kind = stge_kind
+        self.Bilt.parameters_solver = parameters_solver
+        self.Modl = Modelements()
+        self.Modl.Transitions = {}
+
+    def define_transitions_possible(self):
+        # Below: Transition types in their canonical possible order
+
+        # These equations should be used by both solution and simulation code
+
+        # We use an OrderedDict so that the Monte Carlo simulation machinery
+        # will know the order in which they should be executed
+        # (even though the Spyder variable explorer seems, confusingly, to be
+        # unable to show an OrderedDict in its intrinsic order)
+
+        # The name of each equation corresponds to a variable that will be made
+        # and could be preserved in the simulation (or not)
+
+        # Each status will end up transiting only to one subsequent status
+        # ("status" = set of state variables with associated decision problem)
+        # There are multiple possibilities because models may skip many steps
+
+        # Steps can be skipped when the model is one in which nothing happens
+        # in that step, or what happens is so simple that it can be directly
+        # captured in a transition equation
+
+        # For example, with no end of stage shocks, you could go directly from
+        # the "chosen" (after choice) status to the "next_BOP" status
+
+        # Equations that define transitions that affect agent's state
+        transitions_possible = {  # BOP: Beginnning of Problem
+            'BOP_to_choice': {},
+            'choice_to_chosen': {},  # or
+            'choice_to_next_BOP': {},
+            'chosen_to_EOP': {},  # or
+            'chosen_to_next_BOP': {},  # or
+            'chosen_to_next_choice': {},
+            'EOP_to_next_BOP': {},  # or
+            'EOP_to_next_choice': {},  # EOP: End of Problem/Period
```

Comment:
Support for BOP and EOP here:
https://github.com/econ-ark/HARK/pull/1039#issuecomment-895637114